### ISSUE TITLE
Xpand Registration - create new user fields are read only - protected content

### ISSUE THREAD

**Issue by dtbagg** [eXpandFramework/eXpand#421]
**Describe the bug**
After enabling Registration in Model, logout and login.  Register button is visible but when clicked the fields are read only and display protected content.

**Expected behavior**
I expected to be able to enter new user information.

* maybe user error - but seems like a bug.  This was using emailtester demo app.

**Comment by apobekiaris** [comment]
hmm I need more details I cannot reproduce the case. Can you post exact steps on what you did?

**Comment by dtbagg** [comment]
Attached PDF files has steps I took and screen shots of results.  Likely user error on my part instead of a bug, but I can't see what step I am missing.
[protectedcontent.pdf](https://github.com/eXpandFramework/eXpand/files/3234138/protectedcontent.pdf)

**Comment by dtbagg** [comment]
Just now did the same process but set the web app as start up.  I get the same protected content - see image.  
![webappalsoprotected](https://user-images.githubusercontent.com/51164046/58603484-ec081480-825e-11e9-92ae-8c2c849ca33c.JPG)

But I might be on to something, I don't know.  I did see an error - I may not have downloaded something correctly with Nuget - I see this:  1.1.21.2' of package 'Xpand.XAF.Modules.ModelViewInheritance and I think it related to Xpand security.  I will keep trying.  Will post again if I get a clue.

**Comment by dtbagg** [comment]
I got rid of the ModelView error but then thought maybe the issue was with the database.  So as I test I created an empty emailtester.mdb file and modified web.config and tried again.  Same problem, but then I decided to try and add a customer and got a path error. Seems it was looking for an emails folder under \localdata\temp so I created that folder and tried again.  This time no error when adding a customer, but still protected content when trying to register a new user.  Also, same error just trying to change Admin password.  Finally after adding a customer record or two and exiting the record is not actually added - to the database, I don't think I saw an appdata entry.  Modified web.config like this:  <add name="ConnectionString" connectionString="Provider=Microsoft.Jet.OLEDB.4.0;Password=;User ID=Admin;Data Source=C:\inetpub\wwwroot\db\EmailTester.mdb;Mode=Share Deny None;"/>

I had also noted the sql connection string pointing to sqlexpress.  I deleted that and tried again  still seems there is some database permissions issue going on. I can add a customer now, but don't see the data getting created. Just about to give up tonight.  Maybe I will find the problem tomorrow.

**Comment by skybluex3** [comment]
İ guess its about roles.
If you using securitysystemrole then
Set the rolemodule.AllowAnonymousAccess= true

**Comment by apobekiaris** [comment]
>But I might be on to something, I don't know. I did see an error - I may not have downloaded something correctly with Nuget - I see this: 1.1.21.2' of package 'Xpand.XAF.Modules.ModelViewInheritance and I think it related to Xpand security. I will keep trying. Will post again if I get a clue.

see #420 

I did not research the rest yet, will update u again

**Comment by dtbagg** [comment]
> 
> 
> İ guess its about roles.
> If you using securitysystemrole then
> Set the rolemodule.AllowAnonymousAccess= true

I did not see where to change in the model, searched solution and found AllowAnonymousAccess was set to false, set it to true but this didn't seem to make any difference.  Still something is strange when I login with Admin and then it appears I can add a customer record, but the database I thought was being used does not seem to be updated.  I guess I need to look at updater.cs or something else.  I thought this would be so easy but I think I must be missing something that is obvious.

**Comment by apobekiaris** [comment]
try this after on application setupcompleted event
```cs
((SecurityStrategy) SecuritySystem.Instance).AnonymousAllowedTypes.Add(typeof(RegisterUserParameters));
```

**Comment by apobekiaris** [comment]
>AllowAnonymousAccess

this relates somehow to the registration in the sense that when enable the registration is not displayed at startup as the app logins with a pre configure anonymous role but is displayed when you try to login explicitly with the avaiable action in each UI style.

About the tester solutions see #417

**Comment by dtbagg** [comment]
I have had no luck.  I don't have any more time to try right now, but I thank you for your suggestions.  I saw the comment about Demo vs Testing.  May be unrelated but I was just trying to find one sample (maybe that is a better word than demo) for registering a user/email is a bonus.  In git hub I found source with for 19.1 with folder \demos\modules and then email.  This is what I was trying to use. In more recent file on github I didn't find the same folder, instead under \demos\ I only found feature center, xvideo rental and Middle Tier.  Again, it is great to see demo applications running that demonstrate all the cool features in xpand.  But more helpful to some like me would be just simple sample applications that I could test and modify and test again.  ( don't really use easy tester yet, but if a sample compiles and runs then I know it can help me to modify and try and see if I did something to mess it up or if it still works.  

Thanks for your help and patience - I have to move on to other things right now.

**Comment by apobekiaris** [comment]
sorry to hear u spent that much time on this, in previous versions there was a folder modules with Tester solutions. These solution cause missunderstanding like this one. These solutions were there only for the included easytests and not for demo anything and they are now removed.

I attached a short screenshot of the path one should follow for all modules of the main project as per Getting started doc.
1. Install eXpand from the script in https://github.com/eXpandFramework/eXpand/releases so assmeblies will go into GAC and VSIX will be installed which contains the New Project Wizard
2. Create a New ExpandFrameework project from the wizard with authentication in XAF Wizard which will lead to Xpand Security module which apparently will contain something like registration.
3. Enable registration
4. Run and all set.

[2019-05-30_17-01-19.zip](https://github.com/eXpandFramework/eXpand/files/3237365/2019-05-30_17-01-19.zip)

>Again, it is great to see demo applications running that demonstrate all the cool features in xpand. But more helpful to some like me would be just simple sample applications that I could test and modify and test 

there is an ongoing effort for wel document tested modules which are already integrated in this project. You can read more about them at https://github.com/eXpandFramework/DevExpress.XAF

**Comment by dtbagg** [comment]
Well I tried to install again using the scripts so I could have the Xpand project options.  No luck.  Just not meant to be I guess.  See screen shots.
![installerrorswithxpandposhetcsecond](https://user-images.githubusercontent.com/51164046/58660805-4e5f2480-82f4-11e9-9855-b2d6d38d1d25.JPG)
![installerrorswithxpandposhetc](https://user-images.githubusercontent.com/51164046/58660807-4e5f2480-82f4-11e9-9191-b973f99191bf.JPG)

**Comment by apobekiaris** [comment]
hmm this is the uninstall script in the shots you posted which runs only if an installation found and from what I see the assemblies are in the GAC. 

>number of assemblies uninstalled = 1

I also understand that the VSIX failed to install with the script in your system as the uninstall screenshot you posted mentions that the plugin is not installed. 

I need more info to diagnose the case e.g. the `View Install Log` on the bottom left of the VSIX installer during the installation process.

Note that the VSIX exists also as separate download in https://github.com/eXpandFramework/eXpand/releases

**Comment by apobekiaris** [comment]
I did some additional works on the installer related to this case.

1. Remove the VSIXBootStrapper from the UnInstaller and use a manual delete of the installation folders, which is less likely to fail and easier to debug.
2. Create a new Test-Installer-Script pipeline that will run on Azure before each build.
https://github.com/eXpandFramework/Azure-Tasks#test-installer-script

**Comment by dtbagg** [comment]
I am doing better - VSIX not working but I think that may be on my end.  Uninstalling and reinstalling VS to see if that helps. I appreciate your replies.

**Comment by apobekiaris** [comment]
note that the Xpand.Vsix.csproj is a simple standlone project you can open and run from within VS

https://github.com/eXpandFramework/eXpand/tree/master/Xpand.Plugins/Xpand.VSIX

**Comment by dtbagg** [comment]
Thanks, I may try that.  Earlier I found that I had an older version of Xpand installed - I think it was with an MSI installer I don't recall.  I removed it using control panel, then ran the script in PS - may have used Posh Xpand as well but it seemed to work well, but I saw an error message the vsix was not installed something about unsigned, but also saw a VSIX error about dependency graph.  I am now uninstalling and reinstalling VS 2017 (I also found I had still on this computer VS 2012 - very old) so uninstalling that first.  Maybe some cleanup will help.  I will keep advised, but don't think I need to reopen.  Thanks again

**Comment by expand** [comment]
eXpand.lab release [19.1.303.2](https://github.com/eXpandFramework/eXpand.lab/releases/19.1.303.2) includes commit [ExcelImporter: Provide details for AutoImport skipreason in the UI #421](https://github.com/eXpandFramework/eXpand.lab/commit/43989a1e982bd6dfa510e887577b5c873714fab5),[Add Registration parameters to AnonymousAllowedTypes #421](https://github.com/eXpandFramework/eXpand.lab/commit/a8ada0ae0121d2090031d8251093ef581da8d768) that relate to this task. Please test if it addresses the problem. If you use nuget add our LAB [NugetServer](https://xpandnugetserver.azurewebsites.net/nuget) as a nuget package source in VS.

Thanks a lot for your contribution.

**Comment by dtbagg** [comment]
Ok, I had uninstalled old version of MS VS (now just 2017) and I had reinstalled DevExpress,  I still did not see the Xpand Projects so I ran the above from the LAB - then when I went back to VS I still did not have Xpand Project option, but had somehow lost the DevExpress Project types.  I had run the PowerShell script and saw Expand being installed and then the VSIX ran and showed the single VS version.  Seemed to be fine.  Then restarted Visual Studio and now DevExpress or Xpand projects at all and Xpand doesn;t show in the Add/Remove programs or on the start menu.  There is PS script to uninstall in the folder, where installed.  I am about to go back and try again with version 18.2 DevExpress and Xpand.  I have spent a lot of time trying to figure this out, but keep thinking it will be worth it in the end.  I will let you know if I make any progress.

**Comment by apobekiaris** [comment]
just download the VSIX from https://github.com/eXpandFramework/eXpand/releases/download/19.1.303.0/Xpand.VSIX-19.1.303.0.vsix and run it from your explorer. Does it Run? does it finish ok? 

VSIX uninstallation is done from within VS in Tools/Extensions & Updates menu.

The process is as simple as i describe it. It is tests for VS 2017/2019, especially for 2017 the VSIX installation is tested on the Azure cloud.

**Comment by apobekiaris** [comment]
the script in addition to automate the downloads adds the assemblies in the GAC and creates a registry entry so the installation path can be located. It is super simple.

**Comment by dtbagg** [comment]
> 
> 

> VSIX uninstallation is done from within VS in Tools/Extensions & Updates menu.

XPAND itself is not on the Start menu of in the add remove programs menu.  I can (and probably will) run the power shell script to uninstall.
> 
> The process is as simple as i describe it. It is tests for VS 2017/2019, especially for 2017 the VSIX installation is tested on the Azure cloud.

> just download the VSIX from https://github.com/eXpandFramework/eXpand/releases/download/19.1.303.0/Xpand.VSIX-19.1.303.0.vsix and run it from your explorer. Does it Run? does it finish ok?
> 

I am sure any problem is on my end.  But what do you mean my employer? I have used Visual Studio and DevExpress since before I started by current job - more than 6 years ago.  I am working on my own personal computer with Xpand.  Xpand is not installed on my computer for my employer - although I was instrumental in getting DevExpress purchased for my employer and 2 different contracting companies that work for my employer.  I don't know why you ask about my employer?  But I am not a developer on my "real" job.  I have had titles of Database Manager and Project Manager, but am for the most part I am an IT Security Specialist with among many other certifications, CISSP, CISM, CRISC  google them.  So yes, I am sure the process is very simple, and you think I am an idiot.  I will only say over the last 3 years I have on several occasions had our highly paid developers on contract tell my employer this or that task could not be done using DevExpress and I have been able to demonstrate they were wrong.  

I won't bother you again.  You have an excellent add on for DevExpress and any difficulty I am having may well be me of this "personal" computer I have used for a few years and registry likely not kept clean, etc.  If I have another question for you - I *may* ask here again,  right now I would say thank you and I will try not to bother you again over something that is as you say, simple.  good night.


By the way, on your excellent Import Wizard, I believe instead of "Guest" you mean "Guessed".

**Comment by apobekiaris** [comment]
>I am sure any problem is on my end. But what do you mean my employer? 

never said anything about `employer`

>XPAND itself is not on the Start menu of in the add remove programs menu. I can (and probably will) run the power shell script to uninstall.

eXpandFramework was never in the start menu or in the control panel. Manual uninstallation can be achieved from within VS as for any VS plugin

![image](https://user-images.githubusercontent.com/159464/58775184-00387280-85cd-11e9-85cf-7bf4bc17724a.png)

![image](https://user-images.githubusercontent.com/159464/58775204-16dec980-85cd-11e9-8e7e-84935322d45f.png)

btw why u still prefer the ImportWizard instead of the `platform agnostic` `asynchronous` and much richer in features ExcelImporter?
https://github.com/eXpandFramework/eXpand#excelimporter

**Comment by dtbagg** [comment]
Sorry,  You wrote explorer  - I read employer.  When I first tried ExcelImporter I had problems - which you fixed - but I had to use what was working flawlessly at the time.  RIght now I have too many different things going on and not enough time to do any of them correctly.  Lack of sleep, etc.  I will leave this alone.  Thanks

**Comment by dtbagg** [comment]
I was able to get the VSIX working so that when I restarted Visual Studio 2017 (Enterprise Edition) I was able to see the option for a New Xpand Project.  Very nice, like it a lot.  However, after creating a new project with the Xpand Project Wizard, then enabling Registration I get the same Protected Content when trying to register a new user.  I think my next step may be to install Xpand on another computer - my Work computer - I have tried to keep that very similar to that used by our contractors but I may go ahead and install Xpand - then uninstall it.  That way I can convince myself, even though I have cleaned up my development computer quite a bit, that there still may be something about my PC that is causing this issue.  I feel better now that I had a few hours sleep but I don't think I will spend more time trying to make the Registration work.  Apologies again, and if I get lucky I will let you know.

**Comment by dtbagg** [comment]
![stillprotected](https://user-images.githubusercontent.com/51164046/58840155-95ba1d80-8632-11e9-8e81-e484bee1094a.JPG)

**Comment by apobekiaris** [comment]
>that there still may be something about my PC that is causing this issue

incorrect assumption, your pc is fine.

is this a lab build you use? if not then you have to add these lines as I mentioned before.

```cs
public override void Setup(XafApplication application) {
            base.Setup(application);
            application.SetupComplete+=ApplicationOnSetupComplete;
        }

        private void ApplicationOnSetupComplete(object sender, EventArgs e) {
            var securityStrategy = ((SecurityStrategy) SecuritySystem.Instance);
            securityStrategy.AnonymousAllowedTypes.Add(typeof(RegisterUserParameters));
            securityStrategy.AnonymousAllowedTypes.Add(typeof(RestorePasswordParameters));
        }
```

**Comment by dtbagg** [comment]
You were correct.  These lines were missing I added them (but also had to add a couple of "usings") and then the registration fields worked as expected - no more protected content!  I thought I had the version from the LAB installed - but maybe not.  Version of xpand dlls appear to be 19.1.303.0  I added devexpress.expressapp.security and xpand.expressapp.security.registration -

The only thing, unrelated maybe after I closed the app, I saw an error message with this: 
There is already a node with Id 'Solution3XPAND.Module.Solution3XPANDModule'. The node: Application/ModulesList/Solution3XPAND.Module.Solution3XPANDModule. To avoid this error, analyze your custom code or XAFML model differences for possible duplicate nodes creation or follow the www.devexpress.com/kb=S19870 KB article, if you have several classes in different namespaces, but with the same name.

   at DevExpress.ExpressApp.Model.Core.ModelNode.SetId(String value)
   at DevExpress.ExpressApp.Model.Core.ModelNode.SetValueInThisLayer[ValueType](String name, Int32 aspectIndex, ValueType value)
   at DevExpress.ExpressApp.Model.Core.ModelNode.SetValueCore[ValueType](String name, ValueType value, Boolean checkCurrentValue)
   at ModelModule.set_Name(String value)
   at Xpand.Persistent.Base.General.Model.ModuleNodesGenerator.GenerateNodesCore(ModelNode node)
   at DevExpress.ExpressApp.Model.ModelNodesGeneratorBase.GenerateNodes(ModelNode node)
   at DevExpress.ExpressApp.Model.Core.ModelNode.RunNodesGenerator(ModelNodesGeneratorBase generator)
   at DevExpress.ExpressApp.Model.Core.ModelNode.RunNodesGenerator()
   at DevExpress.ExpressApp.Model.Core.ModelNode.EnsureNodes(Boolean alwaysCreate)
   at DevExpress.ExpressApp.Model.Core.ModelNode.GetUnsortedChildren(Boolean inThisLayer)
   at DevExpress.ExpressApp.Model.Core.ModelNodeInfo.GetChildNodes(ModelNode modelNode)
   at DevExpress.ExpressApp.Model.ModelEditorHelper.GetChildNodes(ModelNode modelNode)
   at DevExpress.ExpressApp.Win.Core.ModelEditor.ModelInterfaceAdapter.GetChildrenCore(Object nodeObject)
   at DevExpress.ExpressApp.Win.Core.ModelEditor.ModelInterfaceAdapter.HasChildren(Object nodeObject)
   at DevExpress.ExpressApp.Win.Core.ModelEditor.ExtendModelInterfaceAdapter.HasChildren(Object nodeObject)


I will work more with this sometime tomorrow if I can.  Thanks again for your patience and helpfulness.
